### ISSUE TITLE
The statically-read-axis check withholds unconditionally for a container parameter

### ISSUE THREAD

**Issue by khatchad** [ponder-lab/Hybridize-Functions-Refactoring#914]
Successor to #811 rather than a reopen. That check works as designed for the case it was built for; the case here did not exist when it landed, because nothing produced a container specification entry for a parameter that is also read. #913 produces them, and against such an entry the check withholds unconditionally.

## The Predicate

Both arms short-circuit on any entry that is not a single tensor:

```java
isUnresolvedRead:      if (!(entry instanceof InputSignature.Single single))                       return true;
isRankUnresolvedRead:  if (!(entry instanceof InputSignature.Single single) || dims == null)       return true;
```

The comment on the first records the intent: a container's per-element axes are not attributable to the read, so it declines. The consequence is stronger than conservatism about one axis. **Once a parameter's entry is a sequence, every read against it withholds the whole signature**, whatever the elements are, however concrete. A container parameter that is read at all can never carry a specification.

## Why The Rank Arm Is Wrong Rather Than Merely Cautious

```python
def step(pair):
    x, y = pair
    return tf.reduce_sum(x) + tf.reduce_sum(y)
```

The unpack is a rank-sensitive read of the parameter, and it is the function's *only* static read: no axis is read anywhere. Measured on the witness this came from, the read set is exactly one rank read against parameter 0, with an empty axis-read set.

A declared two-element sequence resolves that read by construction. TensorFlow enforces a declared sequence's structure and raises `ValueError` for a sequence of a different length, so binding a two-name unpack to a two-element specification cannot fail on structure. Executed on the pinned TensorFlow:

| Specification | Result |
| --- | --- |
| `[[TensorSpec((None, 112, 112, 3), tf.float32), TensorSpec((None, None), tf.float32)]]` | runs |
| `[TensorSpec(shape=None, tf.float32)]`, the flat form it replaces | `OperatorNotAllowedInGraphError` |

So the withhold is not protecting against anything here. It costs a signature that runs.

## Negative Control

The finding does not sweep in every rank read, and a fix must not either. A parameter whose entry is a *single* tensor of unknown rank is withheld correctly, through the `dims == null` arm rather than the container arm: there the rank genuinely is unknown, and the same unpack breaks. A witness of that shape exists and its withhold must stay put. Any fix should be verified against both.

## Shape Of A Fix

Attribution rather than a better axis judgment. The check needs to resolve a read against a container entry's elements instead of refusing to look inside, with the rank case answerable from the declared element count alone.

## Sequencing

Not urgent and deliberately queued behind other work on adjacent predicates, so its effect can be attributed on its own. It does not close #888: the structure recovered by #913 is one extent short for a separate reason (#912), and that stands however this is decided.

**Comment by khatchad** [comment]
## Measured On The Full Function, Not A Reduction

The evidence above rests on a reduction that preserves the sole reading site. That is defensible, since the read set says the unpack is the only static read, but it is inference about which parts of the body matter rather than a measurement of the body. Here is the measurement.

The full function was reconstructed and executed: a `MobileNetV2` backbone with pretrained weights behind a two-layer head, a `GradientTape`, `tape.gradient`, `optimizer.apply_gradients`, and a metric update, with the parameter unpacked exactly as the witness unpacks it.

```
undecorated                       2.540102005004883
nested specification              1.4594608545303345
flat specification (the prior)    RAISED OperatorNotAllowedInGraphError
```

The two loss values differ because `apply_gradients` mutates the weights, so the second invocation starts from an already-updated model. What is being measured is that the call completes and returns a loss, which it does.

So the withhold costs a specification that works on the real function. Nothing in the body beyond the unpack interacts with the declared structure, which is what the read set predicted and what this confirms independently of it.
